File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

input = sys.stdin.readline

N, M = map(int, input().split())
MIRO = []
for i in range(N):
    MIRO.append(list(map(int, input().rstrip())))

# 상 하 좌 우
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def bfs(x, y):
    Q = deque([(x, y)])
    while Q:
        x, y = Q.popleft()
        for i in range(4):
            nx = x + dx[i]  # 다음 좌표가 될 수 있는 4가지 경우
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < M and MIRO[nx][ny] == 1:
                Q.append([nx, ny])
                MIRO[nx][ny] = MIRO[x][y] + 1
                if nx == (N - 1) and ny == (M - 1):
                    return MIRO[nx][ny]


print(bfs(0, 0))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
// let input = require('fs').readFileSync('2178.txt').toString().split('\n');
let input = require('fs').readFileSync('dev/stdin').toString().split('\n');

// 행, 열 정의
const [h, w] = input[0].split(' ').map((v) => Number(v));

// 미로 정의
let graph = [];
for (let i = 0; i < h; i++) {
  graph.push(input[i + 1].split('').map((v) => Number(v)));
}

// 좌표 탐색을 위한 dx, dy
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

function bfs(graph, x, y) {
  let queue = [[x, y]];
  while (queue.length) {
    const [x, y] = queue.shift();
    for (let i = 0; i < 4; i++) {
      let nx = x + dx[i];
      let ny = y + dy[i];
      if (nx >= 0 && nx < h && ny >= 0 && ny < w) {
        if (graph[nx][ny] === 1) {
          queue.push([nx, ny]);
          graph[nx][ny] = graph[x][y] + 1;
        }
      }
    }
  }
}

bfs(graph, 0, 0);
console.log(graph[h - 1][w - 1]);
